### PR TITLE
Distinguish between "boundary data" and "mortar data", adding structures

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Distinguishes between "boundary data" (data on an element face that can be projected to a mortar) and "mortar data" (boundary data collected from both sides of a mortar). Used in #1725.

**Note to reviewers:** Because of the renaming it's easier to look at the commits one-by-one.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
@nilsdeppe @kidder could you review this or assign another reviewer while @wthrowe is away? This is the first of a series of PRs that's factored out of #1725.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleMortarData.cpp [hunk 1]
```diff
@@ -28,6 +28,30 @@ SPECTRE_TEST_CASE("Unit.DG.SimpleMortarData", "[Unit][NumericalAlgorithms]") {
...
+                               "[Unit][NumericalAlgorithms]") {
+  ASSERTION_TEST();
+#ifdef SPECTRE_DEBUG
+  dg::SimpleMortarData<size_t, std::string, double> data;
+  data.remote_insert(0, 1.234);
+  data.local_data(0);
+  ERROR("Failed to trigger ASSERT in an assertion test");
+#endif
+}
+// [[OutputRegex, Only have local data at temporal_id]]
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleMortarData.cpp:43]: Blank line.

File: src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
+// Distributed under the MIT License.
```
Comment (wthrowe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp]: Typo in commit message: "proojected".

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
+// Distributed under the MIT License.
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: This file needs to be added to CMakeLists.txt

**Review by wthrowe — CHANGES_REQUESTED** [review]
If I understand correctly, this PR removes (renames) a class called SimpleBoundaryData and then adds a new class with the same name.  I was confused about what was going on for a while.  Perhaps add something to the commit message adding the new version mentioning that something with the same name was removed two commits earlier so that people aren't confused by seeing what appears to be the same class in use before it is added.

**Comment by nilsvu** [comment]
@wthrowe pushed fixups and rephrased the commit messages. Please let me know if the naming is clear now.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp (hunk 3 above)
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: thanks, good catch 👍

**Comment by nilsvu** [comment]
@wthrowe @nilsdeppe could you review when you get the chance? #1586 is very high priority for me and I think this series of PRs that factors the DG and fluxcomm code into more free functions and reusable data structures will be helpful for the subcell communication code :)

**Comment by wthrowe** [comment]
Yes, that's clear now, thanks.

**Comment by nilsvu** [comment]
👍 squashed the fixup because no-one else has reviewed this yet.

**Comment by nilsvu** [comment]
@nilsdeppe could you take a look at this one please?

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@wthrowe please look at

**Comment by kidder** [comment]
reviewed jointly (including @nilsdeppe ) during AEI call

**Review by wthrowe — DISMISSED** [review]

File: src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Variables.hpp"
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp:10]: you're missing some includes

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+#include "Domain/OrientationMapHelpers.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/MortarHelpers.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp"
+#include "NumericalAlgorithms/Spectral/Projection.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+
+struct SomeField : db::SimpleTag {
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: do these need to inherit from SimpleTag

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+struct ExtraData : db::SimpleTag {
+  using type = int;
+};
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.DG.SimpleBoundaryData", "[Unit][NumericalAlgorithms]") {
+  const size_t num_points = 3;
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: why not 5 like we normally use?

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash immediately

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp (hunk 6 above)
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: changed

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp (hunk 5 above)
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_SimpleBoundaryData.cpp]: no they don't 👍

**Comment by nilsvu** [comment]
rebased and squashed. Thanks for reviewing @nilsdeppe @wthrowe @kidder

File: src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp (hunk 4 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp:10]: You still need `pup`

Comment (nilsvu) [reply src/NumericalAlgorithms/DiscontinuousGalerkin/SimpleBoundaryData.hpp:10]: fixed

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**